KEP:  Resource limit-aware scoring plugin

yuanchen8911 opened the pull request:
Propose a resource limit-aware scoring plugin

yuanchen8911 commented:
/Review @Huang-Wei

yuanchen8911 commented:
/cc @Huang-Wei

denkensk commented on kep/206-resource-limit-aware-scoring/README.md:
```diff
@@ -0,0 +1,97 @@
+
+**Default scheduler**
+
+To schedule a new pod pod5 (request: 1, limit: 4), the default scheduler with `LeastAllocated` will choose node1 over node2 as node1 has has a smaller request/allocatable ratio (4/8 vs. 5/8). The resulting node resource status will be
```
```suggestion
To schedule a new pod pod5 (request: 1, limit: 4), the default scheduler with `LeastAllocated` will choose node1 over node2 as node1 has a smaller request/allocatable ratio (4/8 vs. 5/8). The resulting node resource status will be
```

denkensk commented on kep/206-resource-limit-aware-scoring/README.md:
```diff
@@ -0,0 +1,97 @@
+
+1. Calculate each node’s raw score. A node's raw score will be negative for an over-subscription node.
+    
+    - `RawScore = (allocatable - limit) / allocatable`, where limit is the sum of existing pods on the node and the new pod's resource limits.
```
Do we need to think about best effort pods here？

yuanchen8911 replied:
> Do we need to think about best effort pods here？

Very good question. There are two options.

1. Skip best effort pods.

2. Use allocatable as best effort pods's limits. 

BTW, I suspect best effort pods are not allowed in most production environments.

yuanchen8911 replied:
The default can be the node's allocatable.

Huang-Wei commented on kep/206-resource-limit-aware-scoring/README.md:
```diff
@@ -0,0 +1,96 @@
+  - [Implementation History](#implementation-history)
+  - [References](#references)
+<!-- /toc -->
+
```
The template is expected to have a `Summary` section, isn't it?

yuanchen8911 replied:
> The template is expected to have a `Summary` section, isn't it?

Added a summary section.

Huang-Wei commented on kep/206-resource-limit-aware-scoring/README.md:
```diff
@@ -0,0 +1,96 @@
+
+A custom scheduler scoring plugin scores a node by considering existing pods’ resource limits to the node's allocatable ratio on the node. It attempts to “evenly” spread or distribute pod resource limits across nodes in a cluster. 
+
+## Design Details
```
One piece missing in this section (or can be added into the "Implementation" section): how would you like to cache the `limits` of each node? It's not provided by scheduler framework's Snapshot - which has Requested and Allocatable only. So I'd expect to see the details explaining how the `limits` are managed.

yuanchen8911 replied:
> One piece missing in this section (or can be added into the "Implementation" section): how would you like to cache the `limits` of each node? It's not provided by scheduler framework's Snapshot - which has Requested and Allocatable only. So I'd expect to see the details explaining how the `limits` are managed.

Added an implementation section describing the details.  `limits` can be calculated from cached `NodeInfo.Pods (pod.Spec.Containers)`.

yuanchen8911 replied:
> One piece missing in this section (or can be added into the "Implementation" section): how would you like to cache the `limits` of each node? It's not provided by scheduler framework's Snapshot - which has Requested and Allocatable only. So I'd expect to see the details explaining how the `limits` are managed.

It'll be calculated from cached `NodeInfo.Pods`. Add the detail to the proposal. 

   for _, pod := range nodeInfo.Pods {
                for _, container := range pod.Pod.Spec.Containers {
                       ...
                }
  }

Huang-Wei commented on kep/206-resource-limit-aware-scoring/kep.yaml:
```diff
@@ -0,0 +1,13 @@
+title: Resource limit-aware Scoring
+ kep-number: 206
+ authors:
+   - "@yuanchen8911"
```
```suggestion
kep-number: 217
authors:
- "@yuanchen8911"
```

yuanchen8911 commented on kep/206-resource-limit-aware-scoring/README.md:
```diff
@@ -0,0 +1,96 @@
+    
+    - `RawScore = (allocatable - limit) / allocatable`, where limit is the sum of existing pods on the node and the new pod's resource limits.
+    
+    - For multiple resources, the raw score is the weighted sum of all allocatable resources. The resources and their weights are configurable as the plugin arguments.  
```
Yes, it will support all standard resource types plus ScalarResources like the other plugins. The default . Clarified it in the proposal.

yuanchen8911 commented:
@Huang-Wei  @denkensk  could you take another look when you get a chance? Thanks!

Huang-Wei commented on kep/217-resource-limit-aware-scorcing/README.md:
````diff
@@ -0,0 +1,148 @@
+
+```
+// calculateAllocatedLimit returns the allocated resource limits on a node
+ func calculateAllocatedLimit(node *framework.NodeInfo, resource v1.ResourceName) int64 {
````
Make sure the logic is consistent with the upstream one: such as InitContainers, PodOverhead, etc.

yuanchen8911 replied:
Added pod limit calculation detail.

Huang-Wei commented on kep/217-resource-limit-aware-scorcing/README.md:
```diff
@@ -0,0 +1,148 @@
+ 		return limits.Cpu().MilliValue()
+ 	case v1.ResourceMemory:
+ 		return limits.Memory().Value()
+ 	default:
```
Ensure the documented core resources like huge-pages, ephemeral-storage are also handled here. (just be consistent with the upstream logic)

yuanchen8911 replied:
Added ephemeral-storage. Is huge-page part of `ScalarResource`? I don't find an example in the upstream plugins?

Huang-Wei replied:
Ah yes, it's categorized as `ScalarResource`. ([IsScalarResourceName()](https://github.com/kubernetes/kubernetes/blob/48dfa2a55420fc5078e993ac863d566962873927/pkg/scheduler/util/utils.go#L145))

yuanchen8911 replied:
It's then covered. I added it to the description.

yuanchen8911 commented on kep/217-resource-limit-aware-scorcing/README.md:
```diff
@@ -0,0 +1,148 @@
+
+node1’s total resource limit remains the same and node2 has a total resource limit of 9. node2 will have a lower risk of resource contention with pod5 running on it. 
+
+## Terms
```
> Remove it if no terms need to explain.

deleted.

Huang-Wei replied:
It's not deleted...

yuanchen8911 replied:
deleted it now:-).

yuanchen8911 commented on kep/217-resource-limit-aware-scorcing/README.md:
```diff
@@ -0,0 +1,148 @@
+
+- A node's allocatable can be directly obtained from `NodeInfo.Allocatable`. 
+
+- A node's resource limit of all running pods on the node is calculated from `NodeInfo.Pods`.
```
For simplicity,  we will calculate it in each cycle. Is it a concern?

Huang-Wei replied:
I'm not that concerned to get it started in this way. But if we want to graduate into beta, or desire to incorporate as an in-tree plugin, we should make it performant.

For now, please add this performance concern in the `Limitations` section.

yuanchen8911 replied:
> I'm not that concerned to get it started in this way. But if we want to graduate into beta, or desire to incorporate as an in-tree plugin, we should make it performant.
> 
> For now, please add this performance concern in the `Limitations` section.

Added it to the limitation section.

yuanchen8911 commented on kep/217-resource-limit-aware-scorcing/README.md:
````diff
@@ -0,0 +1,148 @@
+ 	return 0
+ }
+```
+- For a best effort pod without specifying resource limits, a default value (e.g., the node's allocatable or configurable value) is used. 
````
Good point. I think we can skip guaranteed pods.

Huang-Wei replied:
> If all pods are guaranteed, it will be equivalent to a weight of 2, but I guess it logically makes more sense by including guaranteed pods.

SG. Please add some statements to explain how Guaranteed pods will be scored.

yuanchen8911 replied:
done.

Huang-Wei commented on kep/217-resource-limit-aware-scorcing/README.md:
````diff
@@ -0,0 +1,178 @@
+- A node's allocatable can be directly obtained from `NodeInfo.Allocatable`. 
+
+- Calcuate the scheduling pods' limit
+```
````
```suggestion
```go
```

Huang-Wei commented on kep/217-resource-limit-aware-scorcing/README.md:
```diff
@@ -0,0 +1,182 @@
+
+## Known Limitations
+
+In the first implementation, the total limit of a node's living pods will be calcauted on demand. The complexity will be O(number of nodes * number of pods per node). The scheduler performance can be affected. To improve the performane, a node's resource limit can be cached and updated upon pod events (add/update/delete).
```
```suggestion
In the first version (alpha), the total limits of a node's pods will be calculated on demand, i.e., on the Score() phase of each scheduling cycle, on each node. Given the scoring plugins will run in parallel, the complexity will be `O(# of pods per node)`.

Once we decide to bump the maturity level to beta, or merge it into the default scheduler as an in-tree plugin, we need to optimize the performance such as caching the node's limits upon receiving pod events (add/update/delete).
```

Huang-Wei reviewed (commented):
Some nits. You can make an update along with squashing the commits.

yuanchen8911 commented:
> squash

done.

yuanchen8911 commented:
@denkensk can you take another look when you get a chance? Thanks!

denkensk commented on kep/217-resource-limit-aware-scorcing/README.md:
```diff
@@ -0,0 +1,182 @@
+   
+1. Calculate each node’s raw score. A node's raw score will be negative for an over-subscription node.
+    
+    - `RawScore = (allocatable - limit) / allocatable`, where limit is the sum of existing pods on the node and the new pod's resource limits.
```
RawScore = ((allocatable - limit) * int64(framework.MaxNodeScore)) / allocatable 
If not multiply larger number like the framework.MaxNodeScore, Decimals will be lost in RawScore.

yuanchen8911 replied:
Made the suggested change.

denkensk reviewed (commented):
Thanks 
Small comments

yuanchen8911 commented:
@Huang-Wei @denkensk Thank you for reviewing the proposal and providing  very helpful feedbacks! Hopefully, I've addressed your comments and concerns.

denkensk commented:
/lgtm
Thanks ^_^
@yuanchen8911

Huang-Wei commented:
/approve

Thanks @yuanchen8911 !
